File: go/grains/grains.go
// Package grains provides utilities to calculate the number of grains of wheat
// on a chessboard given that the number on each square doubles.
package grains

import (
	"errors"
)

// Square calculates how many grains were on a given square.
func Square(number int) (uint64, error) {
	if number < 1 || number > 64 {
		return 0, errors.New("invalid argument: number must be betweeen 1 and 64")
	}
	return 1 << (number - 1), nil
}

// Total calculates the total number of grains on the chessboard.
func Total() uint64 {
	return ^uint64(0)
}

File: fsharp/grains/Grains.fs
module Grains

let square (n: int) : Result<uint64, string> =
    if n < 1 || n > 64 then
        Error "square must be between 1 and 64"
    else
        Ok(uint64 (2.0 ** (float n - 1.0)))

let total: Result<uint64, string> = Ok System.UInt64.MaxValue
